Migrate stock_inventory_hierarchical to v8

clonedagain opened the pull request:
Migrate stock_inventory_hierarchical to v8
# Migration/adaptation to Odoo v8

Use `stock.inventory.location_id` (now it's standard)
Move the "generate inventories" wizard to stock_inventory_hierarchical (drop the field exhaustive)
Use v8 API
Drop the exception `HierarchicalInventoryException` in favor of the new standard exception `ValidationError`
Place the sub-inventories outside the notebook on the main inventory form (the notebook is hidden on draft inventories)
Update the translation template and translation into French.
# Improvements

When creating a hierarchy of inventories using the wizard with the "views only" feature, make sure it matches the view hierarchy. In v7 it would ignore views which were inside internal locations, but this is more common in v8 because a location can't be the source or destination of any stock moves.
Automatically open the sub-inventories when the main inventory is opened. This avoids having to compute useless lines for the main inventory and then re-computing the same lines when opening the sub-inventories
Automatically cancel (set to draft) the sub-inventories when the main inventory is canceled.
Add warnings when clicking the buttons "Start inventory" and "Cancel inventory". The two changes above are a bit invasive, so we must make sure users know what's going to happen.
Speed improvement: defer the computation of `parent_left`/`parent_right` to the end of the wizard.
Add documentation (cf README.rst)
# Coding standards

Clearly define (and strictly enforce) the consitency of the states of the inventories with their sub-inventories.
Rewrite the tests in API v8, improve the coverage. Drop the demo data in favor of a much larger test dataset, automatically generated during the tests.
Adopt the newest OCA standards for licences, README
Bump version to 2.0
Update the screen captures

Commit:
Migrate stock_inventory_hierarchical to v8

Migration/adaptation to Odoo v8
===============================
Use `stock.inventory.location_id` (now it's standard)
Move the "generate inventories" wizard to stock_inventory_hierarchical (drop the field exhaustive)
Use v8 API
Drop the exception `HierarchicalInventoryException` in favor of the new standard exception `ValidationError`
Place the sub-inventories outside the notebook on the main inventory form (the notebook is hidden on draft inventories)
Update the translation template and translation into French.

Improvements
============
When creating a hierarchy of inventories using the wizard with the "views only" feature, make sure it matches the view hierarchy. In v7 it would ignore views which were inside internal locations, but this is more common in v8 because a location can't be the source or destination of any stock moves.
Automatically open the sub-inventories when the main inventory is opened. This avoids having to compute useless lines for the main inventory and then re-computing the same lines when opening the sub-inventories
Automatically cancel (set to draft) the sub-inventories when the main inventory is canceled.
Add warnings when clicking the buttons "Start inventory" and "Cancel inventory". The two changes above are a bit invasive, so we must make sure users know what's going to happen.
Speed improvement: defer the computation of `parent_left`/`parent_right` to the end of the wizard.
Add documentation (cf README.rst)

Coding standards
================
Clearly define (and strictly enforce) the consitency of the states of the inventories with their sub-inventories.
Rewrite the tests in API v8, improve the coverage. Drop the demo data in favor of a much larger test dataset, automatically generated during the tests.
Adopt the newest OCA standards for licences, README
Bump version to 2.0
Update the screen captures

clonedagain commented:
@laetitia-gangloff this PR contains the migration of the first module (parent/child relationship). Can you please review it and maybe test it on copies of your customer's databases?

clonedagain commented:
There seems to be a bad interaction with some other modules in the repo, we're investigating.

clonedagain commented:
Problem fixed, ready for review.

elicoidal commented on stock_inventory_hierarchical/README.rst:
```diff
@@ -0,0 +1,121 @@
+and still keep a clear overview of global inventory's status.
+
+The main inventory will be for a general Location, and the sub-inventories
+should contain sub-Locations.
```
sub-locations

elicoidal commented on stock_inventory_hierarchical/README.rst:
```diff
@@ -0,0 +1,121 @@
+The main inventory will be for a general Location, and the sub-inventories
+should contain sub-Locations.
+
+Odoo will make sure the status of the Inventory and it's Sub-Inventories are
```
inventory / sub-inventories

elicoidal commented on stock_inventory_hierarchical/README.rst:
```diff
@@ -0,0 +1,121 @@
+* sub-inventories will be prepared automatically when you prepare the main
+  inventory
+* you must finish the sub-inventories before you can finish the main inventory
+* canceling a sub-inventory is only possible is the parent inventories are
```
should not it be the other way around?
you can only cancel a parent inventory if all sub-inventories are cancelled?

clonedagain replied:
You're right, we kept this in the code but the "cancel" state is not reachable in the standard, so it does not deserve to be in the README.

elicoidal commented on stock_inventory_hierarchical/README.rst:
```diff
@@ -0,0 +1,121 @@
+
+The most convenient way to create a complete hierarchy of inventories is the
+wizard available in the menu
+"Warehouse > Inventory Control > Generate Inventory"
```
please use enumerated list

elicoidal commented on stock_inventory_hierarchical/README.rst:
```diff
@@ -0,0 +1,121 @@
+Processing the inventories
+--------------------------
+Preparing the main inventory will automatically prepare the sub-inventories.
+All the sub-inventories will set to the same date as the main inventory.
```
All the sub-inventories will have the same date as the main inventory.

elicoidal commented on stock_inventory_hierarchical/__openerp__.py:
```diff
@@ -0,0 +1,19 @@
+    "summary": "Group several Inventory adjustments in a master inventory",
+    "version": "8.0.2.0.0",
+    "depends": ["stock"],
+    "author": u"Numérigraphe,Odoo Community Association (OCA)",
```
I am always confused with the "u": is it really necessary? some says no, some yes...

clonedagain replied:
It is a reasonable precaution because of the accent. Various applications read this file (odoo, the app store, the OCA bots...) and I don't expect all to take unicode for granted.

elicoidal commented on stock_inventory_hierarchical/i18n/fr.po:
```diff
@@ -0,0 +1,196 @@
+#: code:addons/stock_inventory_hierarchical/models/stock_inventory.py:90
+#, python-format
+msgid "Error: You can not create recursive inventories."
+msgstr "Erreur: Vous ne pouvez pas créer d'inventaire récursif"
```
period

elicoidal commented on stock_inventory_hierarchical_location/__init__.py:
```diff
@@ -0,0 +1,22 @@
+# -*- coding: utf-8 -*-
```
Please simplify header

clonedagain replied:
It will be done in another PR when we migrate this module (please cf. my reply regarding  `stock_inventory_hierarchical_location/__openerp__.py`)

elicoidal commented on stock_inventory_hierarchical_location/__openerp__.py:
```diff
@@ -0,0 +1,51 @@
+{
+    "name": "Exhaustive and hierarchical Stock Inventories",
+    "summary": "Extra consistency checks",
+    "version": "7.0.1.1.0",
```
This is a v7 module?

clonedagain replied:
Yes, but this is not the module we're migrating in this PR. This is a side-effect of bringing the history in the repo. We need to bring 3 modules's history into the repo together because they have lots of common commits.

If you'd like to help us make it clearer, can you please review this PR before? https://github.com/OCA/stock-logistics-warehouse/pull/137
This way only the last commit would be shown here.

Commit:
Corrections suggested by Eric Caudal

gurneyalex commented on stock_inventory_hierarchical/wizard/generate_inventory.py:
```diff
@@ -0,0 +1,131 @@
+                               help="If set, only inventory "
+                                    "on view location can be created")
+
+    @api.multi
```
@api.model

gurneyalex commented on stock_inventory_hierarchical/wizard/generate_inventory.py:
```diff
@@ -0,0 +1,131 @@
+            'stock.warehouse')
+        warehouse_ids = self.env['stock.warehouse'].search(
+            [('company_id', '=', company_id)], limit=1)
+        return warehouse_ids and warehouse_ids.lot_stock_id.id or False
```
`return warehouse_ids.lot_stock_id.id` (new api magic)

gurneyalex commented on stock_inventory_hierarchical/wizard/generate_inventory.py:
```diff
@@ -0,0 +1,131 @@
+
+        result = self.env['ir.model.data'].get_object_reference(
+            'stock', 'view_inventory_form')
+        view_id = result and result[1] or False
```
result[1] if result else False

gurneyalex commented on stock_inventory_location/__openerp__.py:
```diff
@@ -0,0 +1,79 @@
+   Inventory Lines.
+ - every good that is not in the Inventory Lines is considered lost, and gets
+   moved out of the stock when you confirm the Inventory.
+""",
```
can you move this out to README.rst?

clonedagain replied:
Will do in an upcoming PR targetting this module specifically. For now it is unported.

gurneyalex commented:
:+1:

Commit:
Corrections suggested by Alexandre Fayolle

clonedagain commented:
Rebased to clean up the diff, and corrected as @gurneyalex suggested.

jgrandguillaume commented:
Hi,

Thanks for the contribs, I've just tested it and works fine. Great module !

:+1: 

Regards,

Joël

Commit:
Fix image path in manifest

Commit:
[FIX] missing cache invalidation after SQL update

Commit:
Don't make the ORM recreate the constraint every time

The ORM does a case-sensitive check for an existing constraint, we used to put an upper-case constraint and PostGreSQL returns a lower-wase name for it.

Commit:
Complement tests and make them more readable

clonedagain commented:
@lbellier and I just fixed a missing cache invalidation here, and added a test which demonstrates it. Reviews are still welcome.

gurneyalex commented:
@clonedagain is this ready for merge or are you still working on it?

clonedagain commented:
I think it's ready for merge. @lbellier and I ported the last related module yesterday 
so I don't expect significant changes now.

lmignon commented:
:+1: Successfully used in production.

clonedagain commented:
@lmignon are you using v8 in production already? there's been quite a bit of refactoring due to the standard having locations in inventories.

lmignon commented:
@clonedagain Oups, we use the V7 version... too fast

lmignon commented:
@clonedagain I've deleted my comments with the approval since it was for the V7.... I've read deeper the code and I'm really impressed by the new way inventories are created and prepared. Even if the code is quite complex and use knowingly some SQL queries, it stay readable. To complete my review I should do some functional tests but I've no time at this moment.

clonedagain commented:
Can we merge this please?
